[Outdated] Implementation of CSE for GT_CNS_INT benefits ARM64.

briansull opened the pull request.

briansull committed:
Mark Div/Mod nodes that will use reciprocal multiply with GTF_DIV_USE_MAGIC

briansull committed:
Update the LclVar ref counts for new CSE LclVars

briansull committed:
Change csdHashKey to INT64, change the CSE hash key to size_t

briansull committed:
Update usesMagicNumberDivision() to handle some additional cases

briansull committed:
Update the gtCost values for constant nodes: GT_CNS_INT, GT_CNS_LNG, GT_CNS_STR
ARM64
      -11528 : System.Private.CoreLib.dasm (-0.10% of base)
252 total methods with Code Size differences (222 improved, 30 regressed), 20696 unchanged.
X64
         886 : System.Private.CoreLib.dasm (0.03% of base)
229 total methods with Code Size differences (16 improved, 213 regressed), 20918 unchanged.

briansull committed:
Fix usesMagicNumberDivision for unsigned divide of MIN_INT
Fixes 4 x86 test failures

briansull committed:
If a tree node has had SetRegNum() called to assign it to a physical register and that node becomes a CSE
then we clear that info on the CSE expression and instead set it on the CSE LclVar.

briansull committed:
As part of the CSE of constants work, once we make a CSE we don't want constant prop to undo the CSE operation
We re-purpose the GTF_DONT_CSE flag to tell assertionprop not to perform constant prop on this CSE value

briansull committed:
When we duplicate a conditional test to change a loop into a bottom tested loop
We also will record that the block has a call if the duplicated code contains a call.

briansull committed:
In invariant hoisting the hoist Expr does not have to computed into a specific register,
so clear the RegNum if it was set in the original expression.

briansull committed:
Change the weights of GT_CNS_INT for ARM64
Add support for COMPLUS_JitDisableConstCSE

briansull committed:
Don't auto retry when using the AltJit as that results in two copies of every method in the output file

briansull committed:
Implementation of CSE for GT_CNS_INT using a single CSE for constant values that differ only in the low 12 bits.

briansull commented:
```
Top method improvements (percentages):
        -128 (-26.89% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int,int,int,int,int,int):int
        -104 (-23.64% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int,int,int,int,int):int
         -64 (-20.25% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int,int):int
         -80 (-20.20% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int,int,int,int):int
         -40 (-17.86% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int):int
         -64 (-17.78% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(int,int,int,int,int):int
        -256 (-16.67% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon):int (2 methods)
         -16 (-16.00% of base) : System.Private.CoreLib.dasm - System.MissingFieldException:.ctor(System.String):this
         -16 (-16.00% of base) : System.Private.CoreLib.dasm - System.MissingMethodException:.ctor(System.String):this
        -132 (-15.71% of base) : System.Private.CoreLib.dasm - System.Numerics.Matrix4x4:CreateLookAt(System.Numerics.Vector3,System.Numerics.Vector3,System.Numerics.Vector3):System.Numerics.Matrix4x4
        -208 (-15.12% of base) : System.Private.CoreLib.dasm - System.HashCode:Combine(System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon,System.__Canon):int (2 methods)
```

tannergooding commented on src/coreclr/src/jit/gentree.cpp:
```diff
@@ -3295,57 +3291,192 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
+                    }
+                    else
+                    {
+                        // Arm64 allows any arbitrary 16-bit constant to be loaded into a register halfword
```
HWIntrinsics recently got some support in lowering where we will generate better code if all inputs to a `NI_Vector128_Create` call are constant.
Is there any handling that should be added here to help account for that?

briansull replied:
No because the Lower phase runs after the CSE phase

briansull committed:
Fix for JitDisableConstCSE() == 3

briansull committed:
In CSE added check for mismatched types on GT_CNS_INT nodes

briansull committed:
Fix assert divMod->markedMagicNumberDivision()

briansull committed:
Propagate the gtFlags for a gtCallAddr node
Const CSE may create an assignment node for the target of a calli: indirect call

briansull committed:
Propagating a constant may create an opportunity to use a magic number division

briansull committed:
Only perform conversion of floating point GT_DIV to an exact reciprocal GT_MUL when fgGlobalMorph is true

briansull committed:
Mark the non-standard virtual stub indirection address arg with GTF_DONT_CSE

briansull committed:
Fix the gtCost GT_CNS_LNG for ARM32

TamarChristinaArm commented on src/coreclr/src/jit/gentree.cpp:
```diff
@@ -3256,17 +3252,44 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
+                    costSz = 9;
+                    costEx = 4;
+                }
+                else if (!codeGen->validImmForInstr(INS_mov, (target_ssize_t)loVal))
```
Should this also check for `INS_mvn`?

tannergooding commented on src/coreclr/src/jit/gentree.cpp:
```diff
@@ -3252,17 +3252,44 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
                 costEx = 3;
                 goto COMMON_CNS;
 
+            case GT_CNS_LNG:
```
Wondering why this is needed since the PR is for Arm64? Isn't `GT_CNS_LNG` never created for 64-bit JIT?

tannergooding replied:
Sorry, I worded this poorly. I meant to say the PR title is implying this is just Arm64 work and should likely be updated if its also touching arm32.

briansull replied:
It current;y enables CSE's of constants across all platforms.
On ARM64 we get the largest benefits..

briansull committed:
Update arm32 GT_CNS costs.
Update JitDisableConstCSE to support various options 1-5

briansull committed:
Changes for config variable JitDisableConstCSE - remove values 4 and 5
More fixes for GT_CNS_INT costs

kunalspathak commented on src/coreclr/src/jit/morph.cpp:
```diff
@@ -2705,6 +2705,8 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
         GenTree* indirectCellAddress = gtNewIconHandleNode(addrValue, GTF_ICON_FTN_ADDR);
         indirectCellAddress->SetRegNum(REG_R2R_INDIRECT_PARAM);
+        // Don't attempt to CSE this constant
+        indirectCellAddress->SetDoNotCSE();
```
It is interesting that we are not CSEing indirect cell addresses.

We can perhaps CSE the following code pattern:
```asm
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
        AA0003EF          mov     x15, x0
       ...
```

```asm
        ...
        9000000B          adrp    x11, [RELOC #0x1de756537c0]
        9100016B          add     x11, x11, #0
        F9400160          ldr     x0, [x11]
                          mov   xR, x0     ; store x0 in some register xR
        D63F0000          blr     x0
        ...
                          mov   x0, xR     ; retrieve xR into x0
        D63F0000          blr     x0
        AA0003EF          mov     x15, x0
       ...
```

With this, we can get an improvement of 8 bytes + 1 elimination of memory access. 
I wrote an analyzer asm to find out how many addresses are CSE candidates and the number is huge. From what I noticed, it would by little over 2MB of size reduction.

`Processed 191816 methods. Found 29246 methods containing 259123 groups.`

Details: [cse-candidates.txt](https://github.com/dotnet/runtime/files/4853383/cse-candidates.txt)

@BruceForstall , @CarolEidt , @AndyAyersMS

kunalspathak replied:
AnalyzeAsm code that I used to find out CSE candidates: https://github.com/dotnet/jitutils/pull/273

briansull replied:
I added this limitation because it was causing an assert failure in LSRA.
I will re-check this to see if it is still happens and add a comment if so.

briansull replied:
In our IR the indirection of r11 is not made explicit until lowering so we can't create a CSE for these indirections at this time:

```
        F9400160          ldr     x0, [x11]
        D63F0000          blr     x0
```

briansull replied:
Actually the indirection of r11 is never materialized in our current IR:

```
Generating: N011 (  2,  8) [000039] -------N----        t39 =    CNS_INT(h) long   0xd1ffab1e ftn REG x11 $140
IN0002:             adrp    x11, [RELOC #0xd1ffab1e]
IN0003:             add     x11, x11, #0
                                                              /--*  t39    long   
Generating: N013 (???,???) [000144] ------------       t144 = *  PUTARG_REG long   REG x11
                                                              /--*  t144   long   arg0 in x11
Generating: N015 ( 16, 11) [000005] --C---------         t5 = *  CALL help r2r_ind ref    HELPER.CORINFO_HELP_READYTORUN_NEW REG x0 $1c0
IN0004:             ldr     x0, [x11]
Call: GCvars=0000000000000000 {}, gcrefRegs=180000 {x19 x20}, byrefRegs=0000 {}
IN0005:             blr     x0
```

briansull replied:
If CSE of Constants are enable for arm32 we will hit an assert when we CSE this address:
(So I will add an #ifdef to prevent this on Arm32)

Assertion failed 'candidates != candidateBit' in 'System.Array:ConvertAll(System.__Canon[],System.Converter`2[__Canon,__Canon]):System.__Canon[]' during 'LSRA build intervals' (IL size 64)

kunalspathak replied:
Sure. Is there a follow-up issue to investigate this ARM failure?

briansull committed:
Allow CSE of the R2R indirect param constant address

kunalspathak commented:
>         stubAddrArg  = gtNewIconHandleNode(addr, GTF_ICON_FTN_ADDR);

should we mark this also as `SetDoNotCSE()`?

---
Refers to: src/coreclr/src/jit/morph.cpp:8397 in 99b1da8. [](commit_id = 99b1da839ecc7903e531417c8cc910580054101e, deletion_comment = False)

briansull committed:
Fix Assertion failed 'isValidGeneralDatasize(size)'

briansull commented:
Sample improvement in **BenchmarksGame.RegexRedux_5:**
We create two CSE's for constants:  cse2 in x24 and cse3 in x19
These are examples of shared offset CSE's as they generate multiple address values that differ by a small constant:

```
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
```



```
; Assembly listing for method BenchmarksGame.RegexRedux_5:Bench(System.IO.TextReader,bool):int
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T14] (  4,  4   )     ref  ->  x19         class-hnd
;  V01 arg1         [V01,T15] (  3,  3   )    bool  ->  x20        
;  V02 loc0         [V02,T01] ( 15, 14.50)     ref  ->  x21         class-hnd exact
;  V03 loc1         [V03,T58] (  2,  1.50)     int  ->  x22         ld-addr-op
;  V04 loc2         [V04,T31] (  3,  2.50)     ref  ->  x23         class-hnd
;  V05 loc3         [V05,T33] (  3,  2   )     ref  ->  x25         class-hnd
;  V06 loc4         [V06,T34] (  3,  2   )     ref  ->  x26         class-hnd
;  V07 loc5         [V07,T35] (  3,  2   )     ref  ->  x27         class-hnd
;  V08 loc6         [V08,T36] (  3,  2   )     ref  ->  x28         class-hnd
;  V09 loc7         [V09,T37] (  3,  2   )     ref  ->  [fp+0x28]   class-hnd
;  V10 loc8         [V10,T38] (  3,  2   )     ref  ->  [fp+0x20]   class-hnd
;  V11 loc9         [V11,T39] (  3,  2   )     ref  ->  [fp+0x18]   class-hnd
;  V12 loc10        [V12,T40] (  3,  2   )     ref  ->  [fp+0x10]   class-hnd
;  V13 loc11        [V13,T41] (  3,  2   )     ref  ->  x19         class-hnd
;  V14 loc12        [V14,T32] (  4,  2   )     int  ->   x0         ld-addr-op
;# V15 OutArgs      [V15    ] (  1,  1   )  lclBlk ( 0) [sp+0x00]   "OutgoingArgSpace"
;  V16 tmp1         [V16,T16] (  2,  4   )     ref  ->  x21         class-hnd exact "NewObj constructor temp"
;  V17 tmp2         [V17,T04] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V18 tmp3         [V18,T05] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V19 tmp4         [V19,T06] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V20 tmp5         [V20,T07] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V21 tmp6         [V21,T08] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V22 tmp7         [V22,T09] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V23 tmp8         [V23,T10] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V24 tmp9         [V24,T11] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V25 tmp10        [V25,T12] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V26 tmp11        [V26,T13] (  4,  8   )     ref  ->   x0         class-hnd exact "NewObj constructor temp"
;  V27 tmp12        [V27,T02] ( 11, 11   )     ref  ->  x20         class-hnd exact "dup spill"
;  V28 tmp13        [V28,T19] (  3,  3   )     ref  ->  x24         class-hnd "impAppendStmt"
;  V29 tmp14        [V29,T20] (  3,  3   )     ref  ->  x20         class-hnd "impAppendStmt"
;  V30 tmp15        [V30,T21] (  3,  3   )     ref  ->  x25         class-hnd "impAppendStmt"
;  V31 tmp16        [V31,T22] (  3,  3   )     ref  ->  x26         class-hnd "impAppendStmt"
;  V32 tmp17        [V32,T23] (  3,  3   )     ref  ->  x20         class-hnd "impAppendStmt"
;  V33 tmp18        [V33,T24] (  3,  3   )     ref  ->  x20         class-hnd "impAppendStmt"
;  V34 tmp19        [V34,T25] (  3,  3   )     ref  ->  x28         class-hnd "impAppendStmt"
;  V35 tmp20        [V35,T26] (  3,  3   )     ref  ->  x27         class-hnd "impAppendStmt"
;  V36 tmp21        [V36,T27] (  3,  3   )     ref  ->  x19         class-hnd "impAppendStmt"
;  V37 tmp22        [V37,T28] (  3,  3   )     ref  ->  x19         class-hnd "impAppendStmt"
;  V38 tmp23        [V38,T42] (  2,  2   )     ref  ->  x20         class-hnd "Local could be aliased or is pinned"
;  V39 tmp24        [V39,T29] (  3,  3   )     ref  ->  x19         class-hnd "impAppendStmt"
;  V40 tmp25        [V40,T17] (  2,  4   )     ref  ->  x19         class-hnd "Inlining Arg"
;* V41 tmp26        [V41    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V42 tmp27        [V42    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V43 tmp28        [V43    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V44 tmp29        [V44    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V45 tmp30        [V45    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V46 tmp31        [V46    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V47 tmp32        [V47    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V48 tmp33        [V48    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V49 tmp34        [V49    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;* V50 tmp35        [V50    ] (  0,  0   )    long  ->  zero-ref    "Inlining Arg"
;  V51 tmp36        [V51,T18] (  2,  4   )     ref  ->   x0         "argument with side effect"
;  V52 tmp37        [V52,T43] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V53 tmp38        [V53,T44] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V54 tmp39        [V54,T45] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V55 tmp40        [V55,T46] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V56 tmp41        [V56,T47] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V57 tmp42        [V57,T48] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V58 tmp43        [V58,T49] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V59 tmp44        [V59,T50] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V60 tmp45        [V60,T51] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V61 tmp46        [V61,T52] (  2,  2   )     ref  ->   x3         "argument with side effect"
;  V62 tmp47        [V62,T53] (  2,  2   )     ref  ->  x21         "argument with side effect"
;  V63 tmp48        [V63,T54] (  2,  2   )     ref  ->  x22         "argument with side effect"
;  V64 tmp49        [V64,T55] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V65 tmp50        [V65,T56] (  2,  2   )     ref  ->   x1         "argument with side effect"
;  V66 cse0         [V66,T57] (  3,  1.50)     ref  ->  x21         "CSE - moderate"
;  V67 cse1         [V67,T30] (  3,  3   )     ref  ->  x19         "CSE - moderate"
;  V68 cse2         [V68,T00] ( 20, 19.50)    long  ->  x24         "CSE - aggressive"
;  V69 cse3         [V69,T03] ( 11, 11   )    long  ->  x19         "CSE - aggressive"
;
; Lcl frame size = 32

G_M21070_IG01:
        A9B87BFD          stp     fp, lr, [sp,#-128]!
        A90353F3          stp     x19, x20, [sp,#48]
        A9045BF5          stp     x21, x22, [sp,#64]
        A90563F7          stp     x23, x24, [sp,#80]
        A9066BF9          stp     x25, x26, [sp,#96]
        A90773FB          stp     x27, x28, [sp,#112]
        910003FD          mov     fp, sp
        AA0003F3          mov     x19, x0
        2A0103F4          mov     w20, w1
						;; bbWeight=1    PerfScore 7.50
G_M21070_IG02:
        D29F8B00          movz    x0, #0xfc58
        F2B11E20          movk    x0, #0x88f1 LSL #16
        F2CFFF20          movk    x0, #0x7ff9 LSL #32
        94000000          bl      CORINFO_HELP_NEWSFAST
        AA0003F5          mov     x21, x0
        AA1303E0          mov     x0, x19
        F9400261          ldr     x1, [x19]
        F9402821          ldr     x1, [x1,#80]
        F9401021          ldr     x1, [x1,#32]
        D63F0020          blr     x1
        910022AE          add     x14, x21, #8
        AA0003EF          mov     x15, x0
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        F94006B3          ldr     x19, [x21,#8]
        B9400A76          ldr     w22, [x19,#8]
        D2863600          movz    x0, #0x31b0
        F2A20000          movk    x0, #0x1000 LSL #16
        F2C03740          movk    x0, #442 LSL #32
        F9400000          ldr     x0, [x0]
        94000000          bl      System.Text.RegularExpressions.RegexCache:GetOrAdd(System.String):System.Text.RegularExpressions.Regex
        D2860C02          movz    x2, #0x3060
        F2A20002          movk    x2, #0x1000 LSL #16
        F2C03742          movk    x2, #442 LSL #32
        F9400042          ldr     x2, [x2]
        AA1303E1          mov     x1, x19
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Text.RegularExpressions.Regex:Replace(System.String,System.String):System.String:this
        910022AE          add     x14, x21, #8
        AA0003EF          mov     x15, x0
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        D289B400          movz    x0, #0x4da0
        F2B11EA0          movk    x0, #0x88f5 LSL #16
        F2CFFF20          movk    x0, #0x7ff9 LSL #32
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        D28C6313          movz    x19, #0x6318
        F2B118F3          movk    x19, #0x88c7 LSL #16
        F2CFFF33          movk    x19, #0x7ff9 LSL #32
        F9000C13          str     x19, [x0,#24]
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[Int32]):System.Threading.Tasks.Task`1[Int32]
        AA0003F7          mov     x23, x0
        D28A5418          movz    x24, #0x52a0
        F2B11EB8          movk    x24, #0x88f5 LSL #16
        F2CFFF38          movk    x24, #0x7ff9 LSL #32
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91002261          add     x1, x19, #8
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA0003F9          mov     x25, x0
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91004261          add     x1, x19, #16
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA0003FA          mov     x26, x0
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91006261          add     x1, x19, #24
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA0003FB          mov     x27, x0
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91008261          add     x1, x19, #32
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA0003FC          mov     x28, x0
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        9100A261          add     x1, x19, #40
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        F90017A0          str     x0, [fp,#40]	// [V09 loc7]
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        9100C261          add     x1, x19, #48
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
						;; bbWeight=1    PerfScore 92.00
G_M21070_IG03:
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        F90013A0          str     x0, [fp,#32]	// [V10 loc8]
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        9100E261          add     x1, x19, #56
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        F9000FA0          str     x0, [fp,#24]	// [V11 loc9]
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91010261          add     x1, x19, #64
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        F9000BA0          str     x0, [fp,#16]	// [V12 loc10]
        AA1803E0          mov     x0, x24
        94000000          bl      CORINFO_HELP_NEWSFAST
        9100200E          add     x14, x0, #8
        AA1503EF          mov     x15, x21
        94000000          bl      CORINFO_HELP_ASSIGN_REF
        91012261          add     x1, x19, #72
        F9000C01          str     x1, [x0,#24]
        AA0003E1          mov     x1, x0
        9108A300          add     x0, x24, #552
        94000000          bl      System.Threading.Tasks.Task:Run(System.Func`1[__Canon]):System.Threading.Tasks.Task`1[__Canon]
        AA0003F3          mov     x19, x0
        72001E9F          tst     w20, #255
        54001120          beq     G_M21070_IG06
						;; bbWeight=1    PerfScore 27.00
G_M21070_IG04:
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F8          mov     x24, x0
        F9400FA0          ldr     x0, [fp,#24]
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1803E0          mov     x0, x24
        F9400302          ldr     x2, [x24]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F4          mov     x20, x0
        AA1903E0          mov     x0, x25
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1403E0          mov     x0, x20
        F9400282          ldr     x2, [x20]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F9          mov     x25, x0
        AA1A03E0          mov     x0, x26
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1903E0          mov     x0, x25
        F9400322          ldr     x2, [x25]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003FA          mov     x26, x0
        F94017A0          ldr     x0, [fp,#40]
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1A03E0          mov     x0, x26
        F9400342          ldr     x2, [x26]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F4          mov     x20, x0
        F94013A0          ldr     x0, [fp,#32]
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1403E0          mov     x0, x20
        F9400282          ldr     x2, [x20]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F4          mov     x20, x0
        AA1C03E0          mov     x0, x28
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1403E0          mov     x0, x20
        F9400282          ldr     x2, [x20]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003FC          mov     x28, x0
        AA1B03E0          mov     x0, x27
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1C03E0          mov     x0, x28
        F9400382          ldr     x2, [x28]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003FB          mov     x27, x0
        AA1303E0          mov     x0, x19
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1B03E0          mov     x0, x27
        F9400362          ldr     x2, [x27]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
						;; bbWeight=0.50 PerfScore 70.25
G_M21070_IG05:
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F3          mov     x19, x0
        F9400BA0          ldr     x0, [fp,#16]
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[__Canon][System.__Canon]:get_Result():System.__Canon:this
        AA0003E1          mov     x1, x0
        AA1303E0          mov     x0, x19
        F9400262          ldr     x2, [x19]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F3          mov     x19, x0
        2A1603E0          mov     w0, w22
        94000000          bl      System.Number:Int32ToDecStr(int):System.String
        AA0003F4          mov     x20, x0
        F94006A0          ldr     x0, [x21,#8]
        B9400800          ldr     w0, [x0,#8]
        D2863701          movz    x1, #0x31b8
        F2A20001          movk    x1, #0x1000 LSL #16
        F2C03741          movk    x1, #442 LSL #32
        F9400035          ldr     x21, [x1]
        AA1503F6          mov     x22, x21
        94000000          bl      System.Number:Int32ToDecStr(int):System.String
        AA0003E3          mov     x3, x0
        AA1503E2          mov     x2, x21
        AA1603E0          mov     x0, x22
        AA1403E1          mov     x1, x20
        94000000          bl      System.String:Concat(System.String,System.String,System.String,System.String):System.String
        AA0003E1          mov     x1, x0
        AA1303E0          mov     x0, x19
        F9400262          ldr     x2, [x19]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        94000000          bl      System.Console:get_Out():System.IO.TextWriter
        AA0003F3          mov     x19, x0
        AA1703E0          mov     x0, x23
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[Int32][System.Int32]:get_Result():int:this
        94000000          bl      System.Number:Int32ToDecStr(int):System.String
        AA0003E1          mov     x1, x0
        AA1303E0          mov     x0, x19
        F9400262          ldr     x2, [x19]
        F9404042          ldr     x2, [x2,#128]
        F9400C42          ldr     x2, [x2,#24]
        D63F0040          blr     x2
        1400002B          b       G_M21070_IG07
						;; bbWeight=0.50 PerfScore 33.50
G_M21070_IG06:
        910AC300          add     x0, x24, #688
        D2800121          mov     x1, #9
        94000000          bl      CORINFO_HELP_NEWARR_1_OBJ
        AA0003F4          mov     x20, x0
        AA1403E0          mov     x0, x20
        F9400FA2          ldr     x2, [fp,#24]	// [V11 loc9]
        52800001          mov     w1, #0
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        AA1903E2          mov     x2, x25
        52800021          mov     w1, #1
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        AA1A03E2          mov     x2, x26
        52800041          mov     w1, #2
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        F94017A2          ldr     x2, [fp,#40]	// [V09 loc7]
        52800061          mov     w1, #3
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        F94013A2          ldr     x2, [fp,#32]	// [V10 loc8]
        52800081          mov     w1, #4
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        AA1C03E2          mov     x2, x28
        528000A1          mov     w1, #5
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        AA1B03E2          mov     x2, x27
        528000C1          mov     w1, #6
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        AA1303E2          mov     x2, x19
        528000E1          mov     w1, #7
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        F9400BA2          ldr     x2, [fp,#16]	// [V12 loc10]
        52800101          mov     w1, #8
        94000000          bl      CORINFO_HELP_ARRADDR_ST
        AA1403E0          mov     x0, x20
        94000000          bl      System.Threading.Tasks.Task:WaitAll(System.Threading.Tasks.Task[])
						;; bbWeight=0.50 PerfScore 16.25
G_M21070_IG07:
        AA1703E0          mov     x0, x23
        B940001F          ldr     wzr, [x0]
        94000000          bl      System.Threading.Tasks.Task`1[Int32][System.Int32]:get_Result():int:this
						;; bbWeight=1    PerfScore 4.50
G_M21070_IG08:
        A94773FB          ldp     x27, x28, [sp,#112]
        A9466BF9          ldp     x25, x26, [sp,#96]
        A94563F7          ldp     x23, x24, [sp,#80]
        A9445BF5          ldp     x21, x22, [sp,#64]
        A94353F3          ldp     x19, x20, [sp,#48]
        A8C87BFD          ldp     fp, lr, [sp],#128
        D65F03C0          ret     lr
						;; bbWeight=1    PerfScore 7.00

; Total bytes of code 1376, prolog size 28, PerfScore 395.60, (MethodHash=1582adb1) for method BenchmarksGame.RegexRedux_5:Bench(System.IO.TextReader,bool):int
<
```

kunalspathak commented on src/coreclr/src/jit/jitconfigvalues.h:
```diff
@@ -286,6 +286,14 @@ CONFIG_INTEGER(JitDisableSimdVN, W("JitDisableSimdVN"), 0) // Default 0, ValueNu
+// Default 0, enable the CSE of Constants, including nearby offsets. (only for ARM64)
+// If 1, disable all the CSE of Constants
+// If 2, enable the CSE of Constants but don't combine with nearby offsets. (only for ARM64)
+// If 3, enable the CSE of Constants including nearby offsets. (all platforms)
```
> (all platforms) [](start = 62, length = 16)

just for completion, add ` (all platforms)`

briansull replied:
Yes, I have disable CSE of constants by default for non-ARM64 platforms.

see optcse.cpp lines 726-742

```
    bool disableConstCSE = false;

    int configValue = JitConfig.JitDisableConstCSE();

    // all platforms - disable CSE of constant values when config is 1
    if (configValue == 1)
    {
        disableConstCSE = true;
    }
#if !defined(TARGET_ARM64)
    // non-ARM64 platforms - also disable CSE of constant values when config is 0 or 2
    if ((configValue == 0) || (configValue == 2))
    {
        disableConstCSE = true;
    }
#endif
```

TamarChristinaArm commented:
That looks awesome @briansull , great improvement!. Unrelated question to this, is `CORINFO_HELP_ARRADDR_ST` void? in which case we can avoid repeatedly setting `x0` before the calls if it knows they're still live.

kunalspathak commented on src/coreclr/src/jit/optcse.cpp:
```diff
@@ -402,10 +402,27 @@ void Compiler::optValnumCSE_Init()
+#endif // TARGET_ARM64
+
+    // All Platforms - also allow to combine with nearby offsets, when config is 3
+    if (configValue == 3)
```
>configValue [](start = 8, length = 11)

are you considering adding a JITStress pipeline or something equivalent to test this config value?

briansull replied:
@TamarChristinaArm  Yes CORINFO_HELP_ARRADDR_ST has no return value, but it follows the normal calling convention, which kills the incoming argument values in x0, w1 and x2.  So they have to be reloaded at each call site.

briansull replied:
I wasn't planning on adding that.  I  want to be able to experiment on enabling this for the other platforms.

briansull commented:
@TamarChristinaArm Yes CORINFO_HELP_ARRADDR_ST has no return value, but it follows the normal calling convention, which kills the incoming argument values in x0, w1 and x2. So they have to be reloaded at each call site.

TamarChristinaArm commented:
> @TamarChristinaArm Yes CORINFO_HELP_ARRADDR_ST has no return value, but it follows the normal calling convention, which kills the incoming argument values in x0, w1 and x2. So they have to be reloaded at each call site.

@briansull Ah ok, I wasn't sure if you guys strictly followed the platform calling convention for internal calls too.

TamarChristinaArm commented:
> > @TamarChristinaArm Yes CORINFO_HELP_ARRADDR_ST has no return value, but it follows the normal calling convention, which kills the incoming argument values in x0, w1 and x2. So they have to be reloaded at each call site.
> 
> @briansull Ah ok, I wasn't sure if you guys strictly followed the platform calling convention for internal calls too.

Ugh.. sorry I should stop responding near bed-time :) it's a argument register, no guarantee it would ever be live after a call. sorry for the noise :)

briansull committed:
Change gtCostEx for 8-byte constants on AMD64 to be 2 so that they are CSE candidates

briansull committed:
Workaround for ARM32 assert when CSE of Consts is enabled

briansull committed:
Add a simple sort to compute the best shared const CSE's to use for the definition, sets csdConstDefValue and csdConstDefVN
Implemented shared consts CSE's with a 16-bit offset for x64, can be enabled using COMPLUS_JitDisableConstCSE=3

briansull commented:
@dotnet/jit-contrib PTAL

briansull committed:
Fix in AssertionProp for JitStress issue with beq_r8 test

briansull commented:
Test including JitStress=1 are all passing, 
 failed one test, due to current known issue:
**pauseonstart test failures #38847**

briansull commented:
Will rebase to resolve conflicts,
I want to add the above tests to the priority zero catagory to increase our coverage of Floating point optimizations

briansull committed:
Added additional Priority 0 test coverage for Floating Point optimizations
Coverage for Floating point CSEs, conditional branch AssertionProp and NaN handling

briansull commented:
@dotnet/jit-contrib PTAL

sandreenko commented on src/coreclr/src/jit/assertionprop.cpp:
```diff
@@ -3336,8 +3337,15 @@ GenTree* Compiler::optAssertionPropGlobal_RelOp(ASSERT_VALARG_TP assertions, Gen
-        // Also set the value number on the relop.
-        if (curAssertion->assertionKind == OAK_EQUAL)
+        // set foldResult to either 0 or 1
+        bool foldResult = assertionKindIsEqual;
```
Is it a correctness fix? Were we getting wrong VN pair for `!=` in the past?

briansull replied:
Yes, this is incorrectly assigning the wrong constant value number.  I found similar code 
It was occasionally happening in the framework dlls and the the beq_r8 test cases.
Our code has two chances to fold the expression, the first one will fold compares with two constant arguments,
which is the common case.  If that doesn't fold then it tries folding using the value numbers and that path was wrong and caused the beq_r8 tests to fail with my CSE changes.

sandreenko commented on src/coreclr/src/jit/assertionprop.cpp:
```diff
@@ -4947,6 +4955,12 @@ GenTree* Compiler::optVNConstantPropOnJTrue(BasicBlock* block, GenTree* test)
 Compiler::fgWalkResult Compiler::optVNConstantPropCurStmt(BasicBlock* block, Statement* stmt, GenTree* tree)
 {
+    // Don't perform const prop on expressions marked with GTF_DONT_CSE
+    if (!tree->CanCSE())
```
was it unprofitable or incorrect to do?

briansull replied:
Doing a constant prop here would replace the CSE LclVar with the original constant.
Essentially undoing the CSE of the constant.

sandreenko commented on src/coreclr/src/jit/compiler.h:
```diff
@@ -6353,6 +6356,19 @@ class Compiler
                                    // number, this will reflect it; otherwise, NoVN.
     };
 
+#if defined(TARGET_XARCH)
```
Should this block live in target.h?

sandreenko commented on src/coreclr/src/jit/gentree.h:
```diff
@@ -2853,6 +2861,19 @@ struct GenTreeOp : public GenTreeUnOp
     }
 
+    // returns true if we will use MagicNumber multiplication for this node.
+    bool usesMagicNumberDivision(Compiler* comp);
```
Nit: we prefer to start function with an upper case unless we create them with a prefix or in an old class that does not follow the current cc, so `UsesMagicNumberDivision/CheckMagicNumberDivision` etc.

briansull replied:
Done as part of #39021

sandreenko commented on src/coreclr/src/jit/gentree.h:
```diff
@@ -922,6 +928,8 @@ struct GenTree
 #define GTF_OVERFLOW                0x10000000 // Supported for: GT_ADD, GT_SUB, GT_MUL and GT_CAST.
                                                // Requires an overflow check. Use gtOverflow(Ex)() to check this flag.
 
+#define GTF_DIV_USE_MAGIC           0x80000000 // GT_DIV -- Uses MagicNumber multiplication to compute this division by a constant
```
I am strongly against introducing any magic to our source base, `Use_Reciprocal_Mul_Or_Shift` or something similar to old `DIV_BY_CONST\DIV_USE_NO_DIV` looks better in the long term for me.

sandreenko replied:
Are commits about this magic division related to implementation of CSE and should be in this PR or could be extracted?

briansull replied:
I added this code to prevent  regressions.   It is a big win to perform a division by a constant using the reciprocal multiplication.  There were cases where we would CSE the constant and replace it with a CSE local variable.

I could extract the part of the change and check it in seperately.

briansull replied:
I factored out all of the prerequisite code changes for this into a new PR

#39021

sandreenko reviewed (changes requested):
I think we should avoid using magic because it is very hard to read and understand.

Nit: I think the review could be faster if you create several PRs put of this: with magic division, correctness fixes and CSE optimizations.

briansull commented:
This PR has been replaced by #39096

stephentoub commented:
> This PR has been replaced by #39096

Can this be closed then?
